File: app/src/main/java/com/example/dictionaryapp/deinflector/Deinflector.java
package com.example.dictionaryapp.deinflector;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Iterator;

public class Deinflector {

    private JSONArray reasons;

    public Deinflector(JSONObject reasons) throws JSONException {
        this.reasons = this.normalizeReasons(reasons);
    }

    public JSONArray deinflect(String source) throws JSONException {
        JSONObject temp = new JSONObject();
        temp.put("source", source);
        temp.put("term", source);
        temp.put("rules", 0);
        temp.put("definitions", new JSONArray());
        temp.put("reasons", new JSONArray());

        JSONArray results = new JSONArray();
        results.put(temp);


        for (int i = 0; i < results.length(); ++i){

            int rules = results.getJSONObject(i).getInt("rules");
            String term = results.getJSONObject(i).getString("term");
            JSONArray reasons = results.getJSONObject(i).getJSONArray("reasons");

            for (int j = 0; j < this.reasons.length(); j++){
                JSONArray currReasonInfo = this.reasons.getJSONArray(j);
                String reason = currReasonInfo.getString(0);
                JSONArray variants = currReasonInfo.getJSONArray(1);

                for (int h = 0; h < variants.length(); h++ ){
                    JSONArray currVariant = variants.getJSONArray(h);
                    String kanaIn = currVariant.getString(0);
                    String kanaOut = currVariant.getString(1);
                    int rulesIn = currVariant.getInt(2);
                    int rulesOut = currVariant.getInt(3);

                    if (
                            (rules != 0 && (rules & rulesIn) == 0) ||
                                    !term.endsWith(kanaIn) ||
                                    (term.length() - kanaIn.length() + kanaOut.length()) <= 0
                    ){
                        continue;
                    }
                    JSONObject infoIGot = new JSONObject();
                    infoIGot.put("source", source);
                    infoIGot.put("term", term.substring(0, term.length() - kanaIn.length()) + kanaOut);
                    infoIGot.put("rules", rulesOut);
                    infoIGot.put("definitions", new JSONArray());
                    JSONArray fullReasons = new JSONArray();
                    fullReasons.put(reason);
                    for (int r = 0; r < reasons.length(); r++){
                        fullReasons.put(reasons.getString(r));
                    }
                    infoIGot.put("reasons", fullReasons);
                    results.put(infoIGot);

                }
            }

        }

        return results;


    }


    public static JSONArray normalizeReasons(JSONObject reasons) throws JSONException {
        JSONArray normalizedReasons = new JSONArray();
        for (Iterator<String> it = reasons.keys(); it.hasNext(); ) {
            String mainKey = it.next();
            JSONArray reasonInfo = reasons.getJSONArray(mainKey);
            JSONArray variants = new JSONArray();

            for (int x = 0; x < reasonInfo.length(); x ++){
                JSONObject currInfo = reasonInfo.getJSONObject(x);
                JSONArray temp = new JSONArray();
                temp.put(currInfo.get("kanaIn"));
                temp.put(currInfo.get("kanaOut"));
                temp.put(Deinflector.rulesToRuleFlags((JSONArray) currInfo.get("rulesIn")));
                temp.put(Deinflector.rulesToRuleFlags((JSONArray) currInfo.get("rulesOut")));

                variants.put(temp);
            }
            JSONArray tempReasonVariants = new JSONArray();
            tempReasonVariants.put(mainKey);
            tempReasonVariants.put(variants);

            normalizedReasons.put(tempReasonVariants);

        }
        ;

        return normalizedReasons;
    }

    public static int rulesToRuleFlags(JSONArray rules) throws JSONException {
        JSONObject ruleTypes = new JSONObject();
        ruleTypes.put("v1", 0b0000001);
        ruleTypes.put("v5", 0b0000010);
        ruleTypes.put("vs", 0b0000100);
        ruleTypes.put("vk", 0b0001000);
        ruleTypes.put("adj-i", 0b0010000);
        ruleTypes.put("iru", 0b0100000);

        int value = 0;
        for (int k = 0; k < rules.length(); k++){
            int ruleBits;
            try {
                ruleBits = ruleTypes.getInt(rules.getString(k));
            }
            catch (JSONException e){
                continue;
            }
            value |= ruleBits;
        }
        return value;
    }


}

File: app/src/main/java/com/kamui/rin/deinflector/Deinflector.kt
package com.kamui.rin.deinflector

import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json

data class ReasonInfo<T>(
    val kanaIn: String,
    val kanaOut: String,
    val rulesIn: T,
    val rulesOut: T
)

data class ReasonEntry<T>(
    val reason: String,
    val information: List<ReasonInfo<T>>
)

class Deinflector(deinflectionText: String) {
    private val reasons = normalizeReasons(deinflectionText)

    fun deinflect(source: String): JSONArray {
        val temp = JSONObject()
        temp.put("source", source)
        temp.put("term", source)
        temp.put("rules", 0)
        temp.put("definitions", JSONArray())
        temp.put("reasons", JSONArray())
        val results = JSONArray()
        results.put(temp)
        for (i in 0 until results.length()) {
            val rules = results.getJSONObject(i).getInt("rules")
            val term = results.getJSONObject(i).getString("term")
            val reasons = results.getJSONObject(i).getJSONArray("reasons")
            for (j in 0 until this.reasons.length()) {
                val currReasonInfo = this.reasons.getJSONArray(j)
                val reason = currReasonInfo.getString(0)
                val variants = currReasonInfo.getJSONArray(1)
                for (h in 0 until variants.length()) {
                    val currVariant = variants.getJSONArray(h)
                    val kanaIn = currVariant.getString(0)
                    val kanaOut = currVariant.getString(1)
                    val rulesIn = currVariant.getInt(2)
                    val rulesOut = currVariant.getInt(3)
                    if (rules != 0 && rules and rulesIn == 0 ||
                        !term.endsWith(kanaIn) || term.length - kanaIn.length + kanaOut.length <= 0
                    ) {
                        continue
                    }
                    val infoIGot = JSONObject()
                    infoIGot.put("source", source)
                    infoIGot.put("term", term.substring(0, term.length - kanaIn.length) + kanaOut)
                    infoIGot.put("rules", rulesOut)
                    infoIGot.put("definitions", JSONArray())
                    val fullReasons = JSONArray()
                    fullReasons.put(reason)
                    for (r in 0 until reasons.length()) {
                        fullReasons.put(reasons.getString(r))
                    }
                    infoIGot.put("reasons", fullReasons)
                    results.put(infoIGot)
                }
            }
        }
        return results
    }

    private fun normalizeReasons(deinflectionText: String): List<ReasonEntry<Int>> {
        val entries = Json.decodeFromString<List<ReasonEntry<List<String>>>>(deinflectionText)
        val normalized: MutableList<ReasonEntry<Int>> = ArrayList()
        for (reason in entries) {
            val variants: List<ReasonInfo<Int>> = reason.information.map { i ->
                ReasonInfo(i.kanaIn, i.kanaOut, rulesToRuleFlags(i.rulesIn), rulesToRuleFlags(i.rulesOut))
            }
            normalized.add(ReasonEntry(reason.reason, variants))
        }
        return normalized
    }

    private fun rulesToRuleFlags(rules: List<String>): Int {
        val ruleTypes = mapOf(
            "v1"    to 0b00000001,
            "v5"    to 0b00000010,
            "vs"    to 0b00000100,
            "vk"    to 0b00001000,
            "adj-i" to 0b00010000,
            "iru"   to 0b00100000
        )
        var value = 0
        for (rule in rules) {
            val bits = ruleTypes[rule] ?: continue
            value = value or bits
        }
        return value
    }
}